### ISSUE TITLE
GoVCD VM creation dont have option for entering bootstrap data

### ISSUE THREAD

**Issue by kannanvr** [vmware/go-vcloud-director#315]
Hi,
  When I am comparing the vmware/govmomi project for VM creation options with vmware/go-vcloud-director, I found that We could not provide the bootstrap data with goVCD VM.

In govmomi, project there is a option while creating the VM

type VirtualMachineConfigSpec struct {
....
   	ExtraConfig                  []BaseOptionValue                      `xml:"extraConfig,omitempty,typeattr"`
....
}

We can set boostrapdata on ExtraConfig .

But In GoVCD VM creation does not have such options on the VM structure...
Is there any other alternative way to provide the bootstrapdata on the VM create option with VCD.
or if not available, when will it be available ?

Thanks,
Kannan V

**Comment by vbauzysvmware** [comment]
Hi @kannanvr 
Could you describe what you want to achieve? Our guess you searching for guest properties or initscript functionality. functions like `SetGuestCustomizationSection` or `SetProductSectionList`

**Comment by kannanvr** [comment]
@vbauzysvmware , In Vsphere Vmware API, we can add cloud init Bootstrap data. So that the cloud init fetch the data from cloud provider and initialize the VM

In the below Github URL , you can see the example cloudinit data ... In a similar way we are using to initalize the Vmware vsphere VM
```
https://github.com/vmware/cloud-init-vmware-guestinfo
```

How to provide the similar info to VCD VM ?

We know that we can write a script for VCD VM initialization. But We wanted to initialize via Cloudinit data on VCD VM... Please support us

**Comment by vbauzysvmware** [comment]
You can find answer here: https://github.com/terraform-providers/terraform-provider-vcd/issues/508
function in govcd used for this: `SetGuestCustomizationSection`

**Comment by kannanvr** [comment]
Thanks @vbauzysvmware 

Just 1 FinalQuestion to confirm...
So the cloudinit data i can set on the structure types.GuestCustomizationSection on the parameter CustomizationScript.. Is that my understanding correct ?

**Comment by Didainius** [comment]
@kannanvr ,
What kind of DataSource does your cloud-init use? Is it `DataSourceOVF`?

**Comment by kannanvr** [comment]
@Didainius , I have OVF and OVA. Both the types of data i need to set the cloud init data.
Can you suggest which is the structure i need to fill ?

**Comment by Didainius** [comment]
@kannanvr ,
DataSourceOVF does not directly relate to your image format. Rather it defines how your `cloud-init` is configured to pick metadata. DataSourceOVF(https://cloudinit.readthedocs.io/en/latest/topics/datasources/ovf.html) is one of the methods.

There was a similar thread in Terraform provider discussion (terraform provider uses this SDK) https://github.com/terraform-providers/terraform-provider-vcd/issues/510

In essence - if you want to set OVF properties using this SDK  you can have a look at how it is done in terraform provider (https://github.com/terraform-providers/terraform-provider-vcd/blob/f274ed12d5de651f1ea985a91055b5a72b13131f/vcd/resource_vcd_vapp_vm.go#L1889-L1912)

Functions `func (vm *VM) SetProductSectionList` and `func (vm *VM) GetProductSectionList()` can be used for it.

**Comment by kannanvr** [comment]
Thanks @Didainius ,Its a clear explaination... Which structure  param supports OVA ?

**Comment by Didainius** [comment]
> Thanks @Didainius ,Its a clear explaination... Which structure param supports OVA ?

Could you rephrase? I didn't understand the question.

**Comment by kannanvr** [comment]
@Didainius @vbauzysvmware 
I have tried setting the cloudinit data through code. But i could not see the cloud init data to be executed within the VM

Following is my code to set the cloud init data

```
	vm, err := vapp.GetVMByName("test", true)
	if err!= nil {
		fmt.Println(err)
		os.Exit(1)
	}
		f, err := os.Open("cloudconifg.yml")
	if err!= nil {
		fmt.Println(err)
		os.Exit(1)
	}
	reader := bufio.NewReader(f)
	content, _ := ioutil.ReadAll(reader)
	encoded := base64.StdEncoding.EncodeToString(content)
	fmt.Println("ENCODED: " + encoded)

	vmProperties := &types.ProductSectionList{
		ProductSection: &types.ProductSection{
			Info:     "Guest Properties",
			Property: []*types.Property{},
		},
	}

	oneProp := &types.Property{
		UserConfigurable: true,
		Type:             "string",
		Key:              "user-data",
		Label:            "user-data",
		Value:            &types.Value{Value: encoded},
	}
	vmProperties.ProductSection.Property = append(vmProperties.ProductSection.Property, oneProp)
	_, err = vm.SetProductSectionList(vmProperties)
	if err != nil {
		fmt.Errorf("error Setting guest properties: %s", err)
	}

```
After executing the above mentioned code, i have just reset my VM. But I could not see this cloud init data to be executed on the VM.
In the above code cloudconifg.yml is a file which contains the cloud init data. 
Following is my test cloudconifg.yml file 

```
#cloud-config
users:
  - name: demo
    groups: sudo
    shell: /bin/bash
    sudo: ['ALL=(ALL) NOPASSWD:ALL']
    ssh-authorized-keys:
      - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDf0q4PyG0doiBQYV7OlOxbRjle026hJPBWD+eKHWuVXIpAiQlSElEBqQn0pOqNJZ3IBCvSLnrdZTUph4czNC4885AArS9NkyM7lK27Oo8RV888jWc8hsx4CD2uNfkuHL+NI5xPB/QT3Um2Zi7GRkIwIgNPN5uqUtXvjgA+i1CS0Ku4ld8vndXvr504jV9BMQoZrXEST3YlriOb8Wf7hYqphVMpF3b+8df96Pxsj0+iZqayS9wFcL8ITPApHi0yVwS8TjxEtI3FDpCbf7Y/DmTGOv49+AWBkFhS2ZwwGTX65L61PDlTSAzL+rPFmHaQBHnsli8U9N6E4XHDEOjbSMRX user@example.com
      - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDcthLR0qW6y1eWtlmgUE/DveL4XCaqK6PQlWzi445v6vgh7emU4R5DmAsz+plWooJL40dDLCwBt9kEcO/vYzKY9DdHnX8dveMTJNU/OJAaoB1fV6ePvTOdQ6F3SlF2uq77xYTOqBiWjqF+KMDeB+dQ+eGyhuI/z/aROFP6pdkRyEikO9YkVMPyomHKFob+ZKPI4t7TwUi7x1rZB1GsKgRoFkkYu7gvGak3jEWazsZEeRxCgHgAV7TDm05VAWCrnX/+RzsQ/1DecwSzsP06DGFWZYjxzthhGTvH/W5+KFyMvyA+tZV4i1XM+CIv/Ma/xahwqzQkIaKUwsldPPu00jRN user@desktop
runcmd:
  - touch /test.txt
  - mkdir -p /root/kannan
  - echo "Helloworld" >/root/kannan/world.txt
```

As per the above conversation, you have mentioned that , we can set cloud init data through SetProductSectionList function. But I am not able to set it. There must be some problem on my code.
Can you please guide me how to set the above cloud init data on VM

**Comment by Didainius** [comment]
@kannanvr,
Do you see these properties in vCD UI after executing your code? Also - does this process work for you if set via UI only? 
The reason I am asking is that the problem can as well be related to VM template as cloud-init is not 100% the same everywhere from my practice

**Comment by kannanvr** [comment]
@Didainius ,
I can see that it is present at the VCD UI as below..

![image](https://user-images.githubusercontent.com/16012974/85251793-d5d78280-b477-11ea-91ba-25ddc9d2a7ca.png)

It is not working if i set via UI..

Do you have any valid VM template of centos-7 with cloud init ?

Please share it with us. So that i can try it out with that template

**Comment by Didainius** [comment]
@kannanvr,
I didn't try with CentOS lately, but here are a few things to check:
* What is the version of cloud-init bundled in youer CentOS image?
* Does it also include DataSourcdOVF? (It is a must to make this concept work)

I usually debug such issues by booting up the OS and using serial console to login. Then - at least in Ubuntu - you can look for log in `/var/log/vmware-imc/toolsDeployPkg.log`, `/var/log/cloud-init.log`, `/var/log/cloud-init-output.log`


> DO you see any issue written the piece of code in this comment ?
> [#315 (comment)](https://github.com/vmware/go-vcloud-director/issues/315#issuecomment-647244520)

Not from the first read - but that is not trivial to spot. What is the lifecycle beyond this code snippet? Is the VM already booted before this code? Do you power it on after setting these values? It wouldn't work if the VM is already runnning. This must happen on first boot. 

Your best bet is to check at cloud-init logs and seeing what happens.

**Comment by kannanvr** [comment]
@Didainius , Thanks a lot. now i am able to set cloud int on Ubuntu.
But I am searching for Centos-7 and 8  image with DataSourcdOVF
My centos image cloud init version is 18.3.52
My centOS image is from Vsphere. Vsphere datasource is vmware-guestinfo. But VCD data source is different....

Suppose if i have a base image of centos-7... How to include DataSourcdOVF ?


Thanks for your help. Now i am able to set on ubuntu. If i able to set the cloud init on centos-7, it will be good.
Thanks,
Kannan V

**Comment by nussera** [comment]
Are there any news regarding this issue?
I'm facing it too. It doesn't matter what I set. 'metadata' or 'guest_properties'. No matter the name "hostName" or "guestinfo.hostName" . I can't read the values.  `vmtoolsd --cmd “info-get guestinfo.hostName”` always returns "No value".

Can I somehow help to fix it? Do you know if there needs something to be configured on vCloud to make it work?

Thanks,
Andreas

**Comment by Didainius** [comment]
@nussera ,
There was quite a troubleshooting and in the end it was guest operating system image. You might find this useful - https://github.com/vmware/terraform-provider-vcd/issues/510#issuecomment-843721455 (It is terraform but this SDK is used behind the scenes)
